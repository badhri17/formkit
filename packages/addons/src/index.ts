/**
 * Common, optional, first-party add on features for FormKit.
 *
 * You can add this package by using `npm install @formkit/addons` or `yarn add @formkit/addons`.
 *
 * @packageDocumentation
 */

export * from './plugins/autoAnimatePlugin'
export * from './plugins/multiStep/multiStepPlugin'
export * from './plugins/floatingLabels/floatingLabelsPlugin'
export * from './plugins/localStoragePlugin'
export * from './plugins/autoHeightTextarea'
export * from './plugins/passwordMeterPlugin'
