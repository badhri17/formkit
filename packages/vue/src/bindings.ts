import {
  reactive,
  computed,
  ref,
  watch,
  markRaw,
  triggerRef,
  nextTick,
  isRef,
  isReactive,
} from 'vue'
import {
  FormKitPlugin,
  FormKitFrameworkContext,
  FormKitMessage,
  createClasses,
  createMessage,
  generateClassList,
  FormKitTypeDefinition,
} from '@formkit/core'
import {
  eq,
  has,
  camel,
  empty,
  undefine,
  cloneAny,
  shallowClone,
} from '@formkit/utils'
import { createObserver } from '@formkit/observer'
import { FormKitPseudoProps } from '@formkit/core'

/**
 * A plugin that creates Vue-specific context object on each given node.
 *
 * @param node - FormKitNode to create the context on.
 *
 * @public
 */
const vueBindings: FormKitPlugin = function vueBindings(node) {
  /**
   * Start a validity counter on all blocking messages.
   */
  node.ledger.count('blocking', (m) => m.blocking)
  const isValid = ref<boolean>(!node.ledger.value('blocking'))
  /**
   * Start an error message counter.
   */
  node.ledger.count('errors', (m) => m.type === 'error')
  const hasErrors = ref<boolean>(!!node.ledger.value('errors'))

  /**
   * Keep track of the first time a Vue tick cycle has passed.
   */
  let hasTicked = false
  nextTick(() => {
    hasTicked = true
  })

  /**
   * All messages with the visibility state set to true.
   */
  const availableMessages = reactive<Record<string, FormKitMessage>>(
    node.store.reduce((store, message) => {
      if (message.visible) {
        store[message.key] = message
      }
      return store
    }, {} as Record<string, FormKitMessage>)
  )
  /**
   * A flag that determines when validation messages should be displayed.
   */
  const validationVisibility = ref<string>(
    node.props.validationVisibility ||
      (node.props.type === 'checkbox' ? 'dirty' : 'blur')
  )
  node.on('prop:validationVisibility', ({ payload }) => {
    validationVisibility.value = payload
  })

  /**
   * Keep track of if this input has ever shown validation errors.
   */
  const hasShownErrors = ref(validationVisibility.value === 'live')

  /**
   * If the input is required or not, this is the only validation rule that
   * needs to be explicitly called out since it powers the aria-required attr.
   */
  const isRequired = ref<boolean>(false)
  const checkForRequired = (parsedRules?: Array<{ name: string }>) => {
    isRequired.value = (parsedRules ?? []).some(
      (rule) => rule.name === 'required'
    )
  }
  checkForRequired(node.props.parsedRules)
  node.on('prop:parsedRules', ({ payload }) => checkForRequired(payload))

  /**
   * An array of unique identifiers that should only be used for iterating
   * inside a synced list.
   */
  const items = ref(node.children.map((child) => child.uid))

  /**
   * The current visibility state of validation messages.
   */
  const validationVisible = computed<boolean>(() => {
    if (!context.state) return false
    if (context.state.submitted) return true
    if (!hasShownErrors.value && !context.state.settled) {
      return false
    }
    switch (validationVisibility.value) {
      case 'live':
        return true
      case 'blur':
        return context.state.blurred
      case 'dirty':
        return context.state.dirty
      default:
        return false
    }
  })

  /**
   * Determines if the input should be considered "invalid" â€” note that this is different than a valid input! A
   * valid input is one where the input is not loading, not pending validation, not unsettled, and passes all
   * validation rules. An invalid input is one whose validation rules are not explicitly not passing, and those rules
   * are visible to the user.
   */
  const isInvalid = computed<boolean>(() => {
    return context.state.failing && validationVisible.value
  })

  /**
   * Determines if the input should be considered "complete".
   */
  const isComplete = computed<boolean>(() => {
    return context && hasValidation.value
      ? isValid.value && !hasErrors.value
      : context.state.dirty && !empty(context.value)
  })

  /**
   * If the input has validation rules or not.
   */
  const hasValidation = ref<boolean>(
    Array.isArray(node.props.parsedRules) && node.props.parsedRules.length > 0
  )
  node.on('prop:parsedRules', ({ payload: rules }) => {
    hasValidation.value = Array.isArray(rules) && rules.length > 0
  })

  /**
   * All messages that are currently on display to an end user. This changes
   * based on the current message type visibility, like errorVisibility.
   */
  const messages = computed<Record<string, FormKitMessage>>(() => {
    const visibleMessages: Record<string, FormKitMessage> = {}
    for (const key in availableMessages) {
      const message = availableMessages[key]
      if (message.type !== 'validation' || validationVisible.value) {
        visibleMessages[key] = message
      }
    }
    return visibleMessages
  })

  /**
   * UI Messages.
   */
  const ui = reactive(
    node.store.reduce((messages, message) => {
      if (message.type === 'ui' && message.visible)
        messages[message.key] = message
      return messages
    }, {} as Record<string, FormKitMessage>)
  )

  const passing = computed<boolean>(() => !context.state.failing)

  /**
   * This is the reactive data object that is provided to all schemas and
   * forms. It is a subset of data in the core node object.
   */
  const cachedClasses = reactive<Record<string, string>>({})
  const classes = new Proxy(cachedClasses as Record<PropertyKey, string>, {
    get(...args) {
      if (!node) return ''
      const [target, property] = args
      let className: string | null = Reflect.get(...args)
      if (!className && typeof property === 'string') {
        if (!has(target, property) && !property.startsWith('__v')) {
          const observedNode = createObserver(node)
          observedNode.watch((node) => {
            const rootClasses =
              typeof node.config.rootClasses === 'function'
                ? node.config.rootClasses(property, node)
                : {}
            const globalConfigClasses = node.config.classes
              ? createClasses(property, node, node.config.classes[property])
              : {}
            const classesPropClasses = createClasses(
              property,
              node,
              node.props[`_${property}Class`]
            )
            const sectionPropClasses = createClasses(
              property,
              node,
              node.props[`${property}Class`]
            )
            className = generateClassList(
              node,
              property,
              rootClasses,
              globalConfigClasses,
              classesPropClasses,
              sectionPropClasses
            )
            target[property] = className ?? ''
          })
        }
      }
      return className
    },
  })

  node.on('prop:rootClasses', () => {
    const keys = Object.keys(cachedClasses)
    for (const key of keys) {
      delete cachedClasses[key]
    }
  })

  const describedBy = computed<string | undefined>(() => {
    if (!node) return undefined
    const describers = []
    if (context.help) {
      describers.push(`help-${node.props.id}`)
    }
    for (const key in messages.value) {
      describers.push(`${node.props.id}-${key}`)
    }
    return describers.length ? describers.join(' ') : undefined
  })

  const value = ref(node.value)
  const _value = ref(node.value)

  const context: FormKitFrameworkContext = reactive({
    _value,
    attrs: node.props.attrs,
    disabled: node.props.disabled,
    describedBy,
    fns: {
      length: (obj: Record<PropertyKey, any>) => Object.keys(obj).length,
      number: (value: any) => Number(value),
      string: (value: any) => String(value),
      json: (value: any) => JSON.stringify(value),
      eq,
    },
    handlers: {
      blur: (e?: Event) => {
        if (!node) return
        node.store.set(
          createMessage({ key: 'blurred', visible: false, value: true })
        )
        if (typeof node.props.attrs.onBlur === 'function') {
          node.props.attrs.onBlur(e)
        }
      },
      touch: () => {
        const doCompare = context.dirtyBehavior === 'compare'
        if (node.store.dirty?.value && !doCompare) return
        const isDirty = !eq(node.props._init, node._value)
        if (!isDirty && !doCompare) return
        node.store.set(
          createMessage({ key: 'dirty', visible: false, value: isDirty })
        )
      },
      DOMInput: (e: Event) => {
        node.input((e.target as HTMLInputElement).value)
        node.emit('dom-input-event', e)
      },
    },
    help: node.props.help,
    id: node.props.id as string,
    items,
    label: node.props.label,
    messages,
    didMount: false,
    node: markRaw(node),
    options: node.props.options,
    defaultMessagePlacement: true,
    slots: node.props.__slots,
    state: {
      blurred: false,
      complete: isComplete,
      dirty: false,
      empty: empty(value),
      submitted: false,
      settled: node.isSettled,
      valid: isValid,
      invalid: isInvalid,
      errors: hasErrors,
      rules: hasValidation,
      validationVisible,
      required: isRequired,
      failing: false,
      passing,
    },
    type: node.props.type,
    family: node.props.family,
    ui,
    value,
    classes,
  })

  /**
   * Ensure the context object is properly configured after booting up.
   */
  node.on('created', () => {
    if (!eq(context.value, node.value)) {
      _value.value = node.value
      value.value = node.value
      triggerRef(value)
      triggerRef(_value)
    }
    ;(async () => {
      await node.settled
      if (node) node.props._init = cloneAny(node.value)
    })()
  })

  /**
   * When the node mounts, set the didMount flag.
   */
  node.on('mounted', () => {
    context.didMount = true
  })

  /**
   * Sets the settled state.
   */
  node.on('settled', ({ payload: isSettled }) => {
    context.state.settled = isSettled
  })

  /**
   * Observes node.props properties explicitly and updates them in the context
   * object.
   * @param observe - Props to observe and register as context data.
   */
  function observeProps(observe: FormKitPseudoProps) {
    const propNames = Array.isArray(observe) ? observe : Object.keys(observe)
    propNames.forEach((prop) => {
      prop = camel(prop)
      if (!has(context, prop)) {
        context[prop] = node.props[prop]
      }
      node.on(`prop:${prop}`, ({ payload }) => {
        context[prop as keyof FormKitFrameworkContext] = payload
      })
    })
  }

  /**
   * We use a node observer to individually observe node props.
   */
  const rootProps = () => {
    const props = [
      '__root',
      'help',
      'label',
      'disabled',
      'options',
      'type',
      'attrs',
      'preserve',
      'preserveErrors',
      'id',
      'dirtyBehavior',
    ]
    const iconPattern = /^[a-zA-Z-]+(?:-icon|Icon)$/
    const matchingProps = Object.keys(node.props).filter((prop) => {
      return iconPattern.test(prop)
    })
    return props.concat(matchingProps)
  }
  observeProps(rootProps())

  /**
   * Once the input is defined, deal with it.
   * @param definition - Type definition.
   */
  function definedAs<V = unknown>(definition: FormKitTypeDefinition<V>) {
    if (definition.props) observeProps(definition.props)
  }

  node.props.definition && definedAs(node.props.definition)

  /**
   * When new props are added to the core node as "props" (ie not attrs) then
   * we automatically need to start tracking them here.
   */
  node.on('added-props', ({ payload }) => observeProps(payload))

  /**
   * Watch for input events from core.
   */
  node.on('input', ({ payload }) => {
    if (node.type !== 'input' && !isRef(payload) && !isReactive(payload)) {
      _value.value = shallowClone(payload)
    } else {
      _value.value = payload
      triggerRef(_value)
    }
  })

  /**
   * Model updates from core. This is the raw value and should emitted as a
   * model update even if the value did not update internally. Why? Because
   * the model that created this event may have not be the same value as our
   * internal value.
   *
   * See test: "emits a modelUpdated event even when the value results in the
   * same value"
   */
  node.on('commitRaw', ({ payload }) => {
    if (node.type !== 'input' && !isRef(payload) && !isReactive(payload)) {
      value.value = _value.value = shallowClone(payload)
    } else {
      value.value = _value.value = payload
      triggerRef(value)
    }
    node.emit('modelUpdated')
  })

  /**
   * Watch for input commits from core.
   */
  node.on('commit', ({ payload }) => {
    // The input is dirty after a value has been input by a user
    if (
      (!context.state.dirty || context.dirtyBehavior === 'compare') &&
      node.isCreated &&
      hasTicked
    ) {
      if (!node.store.validating?.value) {
        context.handlers.touch()
      } else {
        const receipt = node.on('message-removed', ({ payload: message }) => {
          if (message.key === 'validating') {
            context.handlers.touch()
            node.off(receipt)
          }
        })
      }
    }
    if (
      isComplete &&
      node.type === 'input' &&
      hasErrors.value &&
      !undefine(node.props.preserveErrors)
    ) {
      node.store.filter(
        (message) =>
          !(message.type === 'error' && message.meta?.autoClear === true)
      )
    }
    if (node.type === 'list' && node.sync) {
      items.value = node.children.map((child) => child.uid)
    }
    context.state.empty = empty(payload)
  })

  /**
   * Update the local state in response to messages.
   * @param message - A formkit message
   */
  const updateState = async (message: FormKitMessage) => {
    if (
      message.type === 'ui' &&
      message.visible &&
      !message.meta.showAsMessage
    ) {
      ui[message.key] = message
    } else if (message.visible) {
      availableMessages[message.key] = message
    } else if (message.type === 'state') {
      context.state[message.key] = !!message.value
    }
  }

  /**
   * Listen to message events and modify the local message data values.
   */
  node.on('message-added', (e) => updateState(e.payload))
  node.on('message-updated', (e) => updateState(e.payload))
  node.on('message-removed', ({ payload: message }) => {
    delete ui[message.key]
    delete availableMessages[message.key]
    delete context.state[message.key]
  })
  node.on('settled:blocking', () => {
    isValid.value = true
  })
  node.on('unsettled:blocking', () => {
    isValid.value = false
  })
  node.on('settled:errors', () => {
    hasErrors.value = false
  })
  node.on('unsettled:errors', () => {
    hasErrors.value = true
  })

  /**
   * Watch the validation visible prop and set the hasShownErrors state.
   */
  watch(validationVisible, (value) => {
    if (value) {
      hasShownErrors.value = true
    }
  })

  node.context = context

  // The context is complete
  node.emit('context', node, false)

  node.on('destroyed', () => {
    node.context = undefined
    /* @ts-ignore */ // eslint-disable-line
    node = null
  })
}

export default vueBindings
