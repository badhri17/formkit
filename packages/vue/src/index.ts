import type { FormKitRootConfig } from '@formkit/core'
/**
 * The official FormKit/Vue integration. This package is responsible for
 * integrating Vue with FormKit core and other first-party packages.
 *
 *
 * @packageDocumentation
 */

declare global {
  // eslint-disable-next-line no-var
  var __FORMKIT_CONFIGS__: FormKitRootConfig[]
}

/**
 * The useInput composable.
 */
export { useInput } from './composables/useInput'

/**
 * Shorthand for creating inputs with standard FormKit features.
 */
export { createInput } from './composables/createInput'

/**
 * The defineFormKitConfig composable.
 */
export { defineFormKitConfig } from './composables/defineFormKitConfig'

/**
 * The plugin and plugin types.
 */
export * from './plugin'

/**
 * The root FormKit component.
 */
export {
  default as FormKit,
  FormKitComponent,
  FormKitSetupContext,
  Slots,
  parentSymbol,
  componentSymbol,
  getCurrentSchemaNode,
} from './FormKit'

/**
 * The FormKitRoot wrapper component used to provide context to FormKit
 * about whether a FormKit input is booting in a Document or ShadowRoot.
 */
export { FormKitRoot, rootSymbol } from './FormKitRoot'

/**
 * The FormKitKitchenSink component.
 */
export { FormKitKitchenSink } from './FormKitKitchenSink'

/**
 * The FormKitMessages component.
 * @public
 */
export { FormKitMessages } from './FormKitMessages'

/**
 * The FormKitProvider component.
 * @public
 */
export {
  FormKitProvider,
  FormKitLazyProvider,
  useConfig,
} from './FormKitProvider'

/**
 * Exports the useFormKitContext composable.
 */
export {
  useFormKitContext,
  useFormKitContextById,
  useFormKitNodeById,
} from './composables/useContext'

/**
 * The FormKitSummary component.
 * @public
 */
export { FormKitSummary, FormKitSummaryMessage } from './FormKitSummary'

/**
 * The FormKitSchema component.
 */
export {
  FormKitSchema,
  Renderable,
  RenderableList,
  RenderableSlot,
  RenderableSlots,
  FormKitComponentLibrary,
  VirtualNode,
} from './FormKitSchema'

/**
 * The default configuration.
 */
export {
  defaultConfig,
  DefaultConfigOptions,
  PluginConfigs,
} from './defaultConfig'

/**
 * The vue specific FormKit core plugin. This is generally required for all
 * vue based FormKit configurations.
 */
export { default as bindings } from './bindings'

/**
 * A vue component for rendering icons from the FormKit icon registry
 */
export { FormKitIcon } from './FormKitIcon'

/**
 * The counter reset for sequential identifiers.
 */
export { resetCount } from './utilities/resetCount'

/**
 * Export the reset count explicitly
 */
export {
  errorHandler,
  setErrors,
  clearErrors,
  submitForm,
  reset,
} from '@formkit/core'

/**
 * Export the changeLocale function explicitly.
 */
export { changeLocale } from '@formkit/i18n'

/**
 * SSR support for cleanup operations relating to SSR.
 */
export { ssrComplete, onSSRComplete } from './composables/onSSRComplete'
