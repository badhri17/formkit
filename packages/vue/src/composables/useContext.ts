import { getNode, watchRegistry, stopWatch } from '@formkit/core'
import type {
  FormKitFrameworkContext,
  FormKitGroupValue,
  FormKitNode,
} from '@formkit/core'
import { parentSymbol } from '../FormKit'
import { ref, inject, onUnmounted } from 'vue'
import type { Ref } from 'vue'

/**
 * Uses the FormKit context to access the current FormKit context. This must be
 * used in a component that is a child of the FormKit component.
 * @param effect - An optional effect callback to run when the context is available.
 */
export function useFormKitContext<T = FormKitGroupValue>(
  effect?: (context: FormKitFrameworkContext<T>) => void
): Ref<FormKitFrameworkContext<T> | undefined>
/**
 * Allows access to a specific context by address.
 * @param address - An optional address of the context to access.
 * @param effect - An optional effect callback to run when the context is available.
 */
export function useFormKitContext<T = FormKitGroupValue>(
  address?: string,
  effect?: (context: FormKitFrameworkContext<T>) => void
): Ref<FormKitFrameworkContext<T> | undefined>
export function useFormKitContext<T = FormKitGroupValue>(
  addressOrEffect?: string | ((context: FormKitFrameworkContext<T>) => void),
  optionalEffect?: (context: FormKitFrameworkContext<T>) => void
): Ref<FormKitFrameworkContext<T> | undefined> {
  const address =
    typeof addressOrEffect === 'string' ? addressOrEffect : undefined
  const effect =
    typeof addressOrEffect === 'function' ? addressOrEffect : optionalEffect
  const context = ref<FormKitFrameworkContext<T> | undefined>()
  const parentNode = inject(parentSymbol, null)
  if (__DEV__ && !parentNode) {
    console.warn(
      'useFormKitContext must be used as a child of a FormKit component.'
    )
  }
  if (parentNode) {
    if (address) {
      context.value = parentNode.at(address)?.context
      const root = parentNode.at('$root')
      if (root) {
        const receipt = root.on('child.deep', () => {
          const targetNode = parentNode.at(address)
          if (targetNode && targetNode.context !== context.value) {
            context.value = targetNode.context as FormKitFrameworkContext<T>
            if (effect) effect(context.value)
          }
        })
        onUnmounted(() => {
          root.off(receipt)
        })
      }
    } else {
      context.value = parentNode?.context
    }
  }
  if (context.value && effect) effect(context.value)
  return context
}

/**
 * Allows global access to a specific context by id. The target node MUST have
 * an explicitly defined id.
 * @param id - The id of the node to access the context for.
 * @param effect - An effect callback to run when the context is available.
 */
export function useFormKitContextById<T = any>(
  id: string,
  effect?: (context: FormKitFrameworkContext<T>) => void
): Ref<FormKitFrameworkContext<T> | undefined> {
  const context = ref<FormKitFrameworkContext<T> | undefined>()
  const targetNode = getNode<T>(id)
  if (targetNode)
    context.value = targetNode.context as FormKitFrameworkContext<T>
  if (!targetNode) {
    const receipt = watchRegistry(id, ({ payload: node }) => {
      if (node) {
        context.value = node.context as FormKitFrameworkContext<T>
        stopWatch(receipt)
        if (effect) effect(context.value)
      }
    })
  }
  if (context.value && effect) effect(context.value)
  return context
}

/**
 * Fetches a node by id and returns a ref to the node. The node in question
 * must have an explicitly assigned id prop. If the node is not available, the
 * ref will be undefined until the node is available.
 * @param id - The id of the node to access.
 * @param effect - An optional effect callback to run when the node is available.
 * @returns
 */
export function useFormKitNodeById<T>(
  id: string,
  effect?: (node: FormKitNode<T>) => void
): Ref<FormKitNode<T> | undefined> {
  const nodeRef = ref<FormKitNode<T> | undefined>()
  const targetNode = getNode<T>(id)
  if (targetNode) nodeRef.value = targetNode
  if (!targetNode) {
    const receipt = watchRegistry(id, ({ payload: node }) => {
      if (node) {
        nodeRef.value = node
        stopWatch(receipt)
        if (effect) effect(node)
      }
    })
  }
  if (nodeRef.value && effect) effect(nodeRef.value)
  return nodeRef
}
