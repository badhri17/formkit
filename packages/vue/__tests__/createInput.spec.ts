import { mount } from '@vue/test-utils'
import { token } from '@formkit/utils'
import { FormKitNode } from '@formkit/core'
import { happy, sad } from '@formkit/icons'
import { createInput } from '../src/composables/createInput'
import CustomCompositionInput from './mocks/CustomCompositionInput'
import CustomOptionsInput from './mocks/CustomOptionsInput'
import FormKit from '../src/FormKit'
import { plugin } from '../src/plugin'
import { defaultConfig } from '../src'
import { nextTick } from 'vue'
import { describe, expect, it, vi } from 'vitest'

describe('schema based inputs', () => {
  it('automatically has labels and help text', () => {
    const foo = createInput('hello')
    const label = token()
    const help = token()
    const wrapper = mount(FormKit, {
      props: {
        type: foo,
        name: 'custom',
        label,
        help,
      },
      global: {
        plugins: [[plugin, defaultConfig]],
      },
    })
    expect(wrapper.html()).toContain(label)
    expect(wrapper.html()).toContain(help)
  })

  it('can output the value of the node', () => {
    const foo = createInput('$_value')
    const value = token()
    const wrapper = mount(FormKit, {
      props: {
        type: foo,
        name: 'custom',
        value,
      },
      global: {
        plugins: [[plugin, defaultConfig]],
      },
    })
    expect(wrapper.html()).toContain(value)
  })

  it('can update a groups value', async () => {
    const foo = createInput({
      $el: 'input',
      attrs: {
        onInput: '$handlers.DOMInput',
        value: '$_value',
      },
    })
    const value = token()
    const wrapper = mount(
      {
        data() {
          return {
            group: {
              wham: value,
            },
          }
        },
        template: `
      <FormKit type="group" v-model="group">
        <FormKit type="foo" name="wham" />
      </FormKit>
      `,
      },
      {
        global: {
          plugins: [
            [
              plugin,
              defaultConfig({
                inputs: { foo },
              }),
            ],
          ],
        },
      }
    )
    expect(wrapper.find('input').element.value).toBe(value)
    wrapper.setData({ group: { wham: 'bizbaz' } })
    await nextTick()
    expect(wrapper.find('input').element.value).toBe('bizbaz')
  })

  it('can create an input with a fragment as the schema', () => {
    const foo = createInput([{ $el: 'header' }, { $el: 'footer' }])
    const wrapper = mount(FormKit, {
      props: {
        type: foo,
        name: 'custom',
      },
      global: {
        plugins: [[plugin, defaultConfig]],
      },
    })
    expect(wrapper.find('header').exists()).toBe(true)
    expect(wrapper.find('footer').exists()).toBe(true)
  })

  it('has prefix and suffix icon support', () => {
    const foo = createInput('FooBar')
    const wrapper = mount(FormKit, {
      props: {
        type: foo,
        name: 'custom',
        prefixIcon: 'happy',
        suffixIcon: 'sad',
      },
      global: {
        plugins: [[plugin, defaultConfig({ icons: { happy, sad } })]],
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('vue component inputs', () => {
  it('can use a vue component with composition api', async () => {
    const box = createInput(CustomCompositionInput)
    const wrapper = mount(
      {
        data() {
          return {
            box: true,
          }
        },
        template: `<FormKit v-model="box" type="box" />`,
      },
      {
        global: {
          plugins: [
            [
              plugin,
              defaultConfig({
                inputs: { box },
              }),
            ],
          ],
        },
      }
    )
    expect(wrapper.find('input').element.checked).toBe(true)
    wrapper.find('input').element.checked = false
    wrapper.find('input').trigger('input')
    await new Promise((r) => setTimeout(r, 22))
    expect(wrapper.vm.box).toBe(false)
  })

  it('can use a vue component with options api', async () => {
    const box = createInput(CustomOptionsInput)
    const wrapper = mount(
      {
        data() {
          return {
            box: true,
          }
        },
        template: `<FormKit v-model="box" type="box" />`,
      },
      {
        global: {
          plugins: [
            [
              plugin,
              defaultConfig({
                inputs: { box },
              }),
            ],
          ],
        },
      }
    )
    expect(wrapper.find('input').element.checked).toBe(true)
    wrapper.find('input').element.checked = false
    wrapper.find('input').trigger('input')
    await new Promise((r) => setTimeout(r, 22))
    expect(wrapper.vm.box).toBe(false)
  })
})

describe('custom input behaviors', () => {
  it('does not emit prop:{property} events for input props', async () => {
    const pseudoPropEvent = vi.fn()
    const nativePropEvent = vi.fn()
    const input = createInput('test input', {
      props: ['bizBaz'],
      features: [
        (node: FormKitNode) => {
          node.on('prop:bizBaz', pseudoPropEvent)
          node.on('prop:delay', nativePropEvent)
        },
      ],
    })
    mount(FormKit, {
      props: {
        type: input,
        bizBaz: 'hello',
        delay: 10,
      },
      global: {
        plugins: [[plugin, defaultConfig]],
      },
    })
    await nextTick()
    expect(nativePropEvent).toHaveBeenCalledTimes(0)
    expect(pseudoPropEvent).toHaveBeenCalledTimes(0)
  })
})

describe('schemaMemoKey', () => {
  it('automatically applies a schema memo key if there is none', async () => {
    const componentA = () => 'Hello world'
    const componentB = () => 'Hello mars'
    const text = createInput(componentA)
    const password = createInput(componentB)
    const wrapper = mount(
      {
        template: `<FormKit type="text" /><FormKit type="password" />`,
      },
      {
        global: {
          plugins: [[plugin, defaultConfig({ inputs: { text, password } })]],
        },
      }
    )
    await nextTick()
    expect(wrapper.text()).toBe('Hello worldHello mars')
  })
})

describe('createInput sectionSchema', () => {
  it('can create an input without a label', () => {
    const foo = createInput(
      'hello',
      {},
      {
        label: null,
      }
    )
    const wrapper = mount(FormKit, {
      props: {
        type: foo,
        name: 'custom',
        label: 'Fizzbuzz',
      },
      global: {
        plugins: [[plugin, defaultConfig]],
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
