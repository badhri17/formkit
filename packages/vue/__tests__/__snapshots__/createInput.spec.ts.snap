// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createInput sectionSchema > can create an input without a label 1`] = `
"<div class=\\"formkit-outer\\" data-empty=\\"true\\">
  <div class=\\"formkit-wrapper\\">
    <!---->
    <div class=\\"formkit-inner\\">
      <!---->
      <!---->hello
      <!---->
      <!---->
    </div>
  </div>
  <!---->
  <!---->
</div>"
`;

exports[`schema based inputs > has prefix and suffix icon support 1`] = `
"<div class=\\"formkit-outer\\" data-empty=\\"true\\" data-prefix-icon=\\"true\\" data-suffix-icon=\\"true\\">
  <div class=\\"formkit-wrapper\\">
    <!---->
    <div class=\\"formkit-inner\\"><span class=\\"formkit-prefix-icon formkit-icon\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 16 16\\"><circle cx=\\"6\\" cy=\\"6\\" r=\\"1\\" fill=\\"currentColor\\"></circle><circle cx=\\"10\\" cy=\\"6\\" r=\\"1\\" fill=\\"currentColor\\"></circle><path d=\\"M8,15c-3.86,0-7-3.14-7-7S4.14,1,8,1s7,3.14,7,7-3.14,7-7,7Zm0-13c-3.31,0-6,2.69-6,6s2.69,6,6,6,6-2.69,6-6-2.69-6-6-6Z\\" fill=\\"currentColor\\"></path><path d=\\"M8,11.5c-1.48,0-2.81-.94-3.3-2.33-.09-.26,.04-.55,.3-.64,.26-.09,.55,.04,.64,.3,.35,1,1.3,1.67,2.36,1.67s2.01-.67,2.36-1.67c.09-.26,.38-.4,.64-.3,.26,.09,.4,.38,.3,.64-.49,1.4-1.82,2.33-3.3,2.33Z\\" fill=\\"currentColor\\"></path></svg></span>
      <!---->FooBar
      <!----><span class=\\"formkit-suffix-icon formkit-icon\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 16 16\\"><circle cx=\\"6\\" cy=\\"6\\" r=\\"1\\" fill=\\"currentColor\\"></circle><circle cx=\\"10\\" cy=\\"6\\" r=\\"1\\" fill=\\"currentColor\\"></circle><path d=\\"M8,15c-3.86,0-7-3.14-7-7S4.14,1,8,1s7,3.14,7,7-3.14,7-7,7Zm0-13c-3.31,0-6,2.69-6,6s2.69,6,6,6,6-2.69,6-6-2.69-6-6-6Z\\" fill=\\"currentColor\\"></path><path d=\\"M10.83,11.5c-.21,0-.4-.13-.47-.33-.35-1-1.3-1.67-2.36-1.67s-2.01,.67-2.36,1.67c-.09,.26-.38,.4-.64,.3-.26-.09-.4-.38-.3-.64,.49-1.4,1.82-2.33,3.3-2.33s2.81,.94,3.3,2.33c.09,.26-.04,.55-.3,.64-.06,.02-.11,.03-.17,.03Z\\" fill=\\"currentColor\\"></path></svg></span>
    </div>
  </div>
  <!---->
  <!---->
</div>"
`;
